### PR TITLE
update user profile

### LINKED ISSUE #35: Feat: User Profile Update
---
### Description

As a user (vendor or buyer), I want to be able to update my profile information, such as name, email, or other relevant details, so that my account information remains up-to-date.

### Acceptance Criteria

- [x] Implement a user profile update API endpoint
- [x] Validate user input and sanitize data
- [x] Update user profile information in the database
- [x] Implement appropriate error handling and responses
- [x] Implement input sanitization

### Developer Notes

- Define the API endpoint and request/response payloads for user profile updates
- Implement input validation and sanitization measures
- Update the user profile information in the database
- Handle edge cases, such as email uniqueness or required fields
- Write unit tests to cover different scenarios and edge cases
- Write clear API documentation for the resulting endpoint

### Related Issues

- #28 

### Definition of Done

- User profile update API endpoint is implemented and functional
- User input is validated, and data is sanitized
- User profile information is updated in the database
- Appropriate error handling and responses are implemented
- Input sanitization measures are in place
- Unit tests are written and passing
- Code adheres to coding standards and best practices
- API documentation for the resulting endpoint is completed

### PR DESCRIPTION
**Pull request by Dawaic6**
#### Implemented Features:
-User Profile Update API Endpoint: Implement an API endpoint /api/v1/updateProfile for users to update their profile information.
-Error Handling and Responses: Implement appropriate error handling mechanisms to handle edge cases and provide informative responses to users.

#### Definition of Done:

- acceptance criteria .
- Unit tests .
- API documentation is included.

Related Issues:
#28

### To use the Code ###
- **Clone this repository**:
  git clone https://github.com/atlp-rwanda/dynamites-ecomm-be.git
- **Run**:
`npm run dev`
- **Test-covarage**:
`npm run test:ci`
- **Test**:
`npm run test`
### Screenshot
-**for test**
image
-**for swagger**
image

### CODE REVIEW

**Review by ambroisegithub — COMMENTED** [review]
It look good for me!!!

**Review by wayneleon1 — DISMISSED** [review]
good job

**Review by 13XAVI — DISMISSED** [review]
Great

**Review by bertrandshema — DISMISSED** [review]
It looks good.

File: src/controller/userController.ts [hunk 1]
```diff
@@ -218,8 +228,52 @@ export const Login = async (req: Request, res: Response) => {
...
+
+if (errors.length > 0) {
+  return res.status(400).json({ errors });
+}
+await userRepository.save(user);
+
+return res.status(201).json({ message: 'User updated successfully' });
+} catch (error) {
+return res.status(500).json({ error: 'Internal server error'});
+}
```
Comment (Habinezajanvier) [inline src/controller/userController.ts]: Please, let's use the middleware with with try catch, to prevent repeating ourselves

File: src/routes/userRoutes.ts [hunk 2]
```diff
@@ -19,5 +20,6 @@ route.post('/login',Login)
 route.get('/all-users', getAllUsers);
 route.post('/verify2FA/:userId', verify2FA);
 
+route.put('/updateProfile/:id',updateProfile);
```
Comment (Habinezajanvier) [inline src/routes/userRoutes.ts]: Remember to embed your functionality into the errorHandler middleware

File: src/controller/userController.ts [hunk 3]
```diff
@@ -218,8 +228,52 @@ export const Login = async (req: Request, res: Response) => {
...
+
+    if (emailExists) {
+      return res.status(400).json({ error: 'Email is already taken' });
+    }
+
+user.email = email;
+}
+if (newPassword) {
+  const hashedPassword = await bcrypt.hash(newPassword, 10); 
+  user.password = hashedPassword;
```
Comment (Habinezajanvier) [inline src/controller/userController.ts]: Update other fields but not the password please

File: src/controller/userController.ts [hunk 4]
```diff
@@ -218,8 +228,52 @@ export const Login = async (req: Request, res: Response) => {
...
 }
+const errors = await validate(user);
 
+
+if (errors.length > 0) {
+  return res.status(400).json({ errors });
+}
```
Comment (Habinezajanvier) [inline src/controller/userController.ts]: You have to validate your request body before anything else please, and please separate concerns, can you try to put validations aside to keep our controller clean

**Review by Habinezajanvier — CHANGES_REQUESTED** [review]

**Review by wayneleon1 — APPROVED** [review]
that's look good

**Review by bertrandshema — APPROVED** [review]
This is okay with me

**Review by Habinezajanvier — APPROVED** [review]
LGTM
